google/cloud/spanner/internal/transaction_impl.h

```diff
@@ -49,7 +49,8 @@ class TransactionImpl {
   // passed TransactionSelector in its Client::Read()/Client::ExecuteSql()
   // call. If initially selector.has_begin(), and the operation successfully
```

Comment:
Not for this PR, but the API is slightly weird.  What if `Functor` had this prototype: `pair<T,std::string>(TransactionSelector const&, std::int64_t)`?  Or maybe even this `T(TransactionSelector const&, std::int64_t, std::string&)` so the implementation of `Visit` is in charge of calling `set_id`?

Reply:
The visitor functor often doesn't need to set the transaction id (it may be a single use transaction, or a single-operation commit, or the typical case of the id already being bound).  So it is difficult for the `Visit` implementation to know when to do so, and building a protocol for the functor to tell it when seems much more complicated than just leaving it to the functor.